internal/database/sqlcommon/tokenaccount_sql.go

```diff
@@ -0,0 +1,167 @@
+// Copyright © 2021 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sqlcommon
+
+import (
+	"context"
+	"database/sql"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/hyperledger-labs/firefly/internal/i18n"
+	"github.com/hyperledger-labs/firefly/internal/log"
+	"github.com/hyperledger-labs/firefly/pkg/database"
+	"github.com/hyperledger-labs/firefly/pkg/fftypes"
+)
+
+var (
+	tokenAccountColumns = []string{
+		"pool_id",
+		"namespace",
+		"token_index",
+		"identity",
+		"balance",
+		"hash",
+	}
+	tokenAccountFilterFieldMap = map[string]string{
+		"poolid":     "pool_id",
+		"tokenindex": "token_index",
+	}
+)
+
+func (s *SQLCommon) UpsertTokenAccount(ctx context.Context, account *fftypes.TokenAccount) (err error) {
+	ctx, tx, autoCommit, err := s.beginOrUseTx(ctx)
+	if err != nil {
+		return err
+	}
+	defer s.rollbackTx(ctx, tx, autoCommit)
+
+	rows, _, err := s.queryTx(ctx, tx,
+		sq.Select("seq").
+			From("tokenaccount").
+			Where(sq.Eq{"hash": account.Hash}),
+	)
+	if err != nil {
+		return err
+	}
+	existing := rows.Next()
+	rows.Close()
+
+	if existing {
+		if err = s.updateTx(ctx, tx,
+			sq.Update("tokenaccount").
+				Set("pool_id", account.Identifier.PoolID).
+				Set("namespace", account.Identifier.Namespace).
+				Set("token_index", account.Identifier.TokenIndex).
+				Set("identity", account.Identifier.Identity).
+				Set("balance", account.Balance).
+				Where(sq.Eq{"hash": account.Hash}),
```

Comment:
... note to self, that I need to understand the significance of the hash (and why it's the key here) - expecting to get there through the rest of the review

Reply:
Mainly because I felt "accounts" were most analogous to "groups", and I followed that as a template. That is, accounts don't have any specific point of creation, they just "exist" for a particular combo of token pool + identity. They're assumed to have zero balance until a transfer/mint causes them to go positive, but since there is no actual creation event, it felt wrong to assign them a UUID - hence the use of a hash instead.

Certainly open to other routes if that doesn't seem right, but that summarizes my line of thought.